java/source/flight.rst

```diff
@@ -0,0 +1,1061 @@
+        private Long rows;
+        public DataInMemory(List<ArrowRecordBatch> listArrowRecordBatch, Schema schema, Long rows) {
+            this.listArrowRecordBatch = listArrowRecordBatch;
+            this.schema = schema;
+            this.rows = rows;
+        }
+        public List<ArrowRecordBatch> getListArrowRecordBatch() {
+            return listArrowRecordBatch;
+        }
+        public Schema getSchema() {
+            return schema;
+        }
+        public Long getRows() {
+            return rows;
+        }
+    }
+
+    // Server
+    Location location = Location.forGrpcInsecure("0.0.0.0", 33333);
+    Map<FlightDescriptor, DataInMemory> dataInMemory = new HashMap<>();
+    Map<String, DataInMemory> mapPojoFlightDataInMemory = new HashMap<>();
+    List<ArrowRecordBatch> listArrowRecordBatch = new ArrayList<>();
+    try (RootAllocator allocator = new RootAllocator(Long.MAX_VALUE)){
+        FlightServer flightServer = FlightServer.builder(allocator, location, new NoOpFlightProducer(){
+            @Override
+            public Runnable acceptPut(CallContext context, FlightStream flightStream, StreamListener<PutResult> ackStream) {
+                return () -> {
+                    long rows = 0;
+                    while (flightStream.next()) {
+                        VectorUnloader unloader = new VectorUnloader(flightStream.getRoot());
+                        try (final ArrowRecordBatch arb = unloader.getRecordBatch()) {
+                            // Retain data information
+                            listArrowRecordBatch.add(arb);
+                            rows = rows + flightStream.getRoot().getRowCount();
+                        }
+                    }
+                    long finalRows = rows;
+                    DataInMemory pojoFlightDataInMemory = new DataInMemory(listArrowRecordBatch, flightStream.getSchema(), finalRows);
+                    dataInMemory.put(flightStream.getDescriptor(), pojoFlightDataInMemory);
+                    ackStream.onCompleted();
+                };
+            }
+
+            @Override
+            public void getStream(CallContext context, Ticket ticket, ServerStreamListener listener) {
+                FlightDescriptor flightDescriptor = FlightDescriptor.path(new String(ticket.getBytes(), StandardCharsets.UTF_8)); // Recover data for key configured
+                if(dataInMemory.containsKey(flightDescriptor)){
+                    VectorSchemaRoot vectorSchemaRoot = VectorSchemaRoot.create(dataInMemory.get(flightDescriptor).getSchema(), allocator);
+                    listener.start(vectorSchemaRoot);
+                    for(ArrowRecordBatch arrowRecordBatch : dataInMemory.get(flightDescriptor).getListArrowRecordBatch()){
+                        vectorSchemaRoot.allocateNew();
+                        VectorLoader loader = new VectorLoader(vectorSchemaRoot);
+                        loader.load(arrowRecordBatch.cloneWithTransfer(allocator));
```

Comment:
Similarly - do we have to clone before loading? That'll copy data but we're not mutating anything.

Reply:
I could suggest this is needed for transferOwnership. If I use `loader.load(arrowRecordBatch)` instead of `loader.load(arrowRecordBatch.cloneWithTransfer(allocator))` I see error `already closed`

Reply:
I think you have to increment the refcount of the batch manually

Reply:
Let me test that

Reply:
Reviewing [Stream example](https://github.com/apache/arrow/blob/master/java/flight/flight-core/src/main/java/org/apache/arrow/flight/example/Stream.java#L163) I see they also use cloneWithTransfer

Probably this is needed by [cloneWithTransfer](https://github.com/apache/arrow/blob/master/java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java#L162:L168) method description mention:

Create a new ArrowRecordBatch which has the same information as this batch but whose buffers are owned by that Allocator.

Reply:
I think incrementing the refcount is sufficient since we are all using the same allocator here but I guess it doesn't matter too much (this will mean we have to copy all data each time though).